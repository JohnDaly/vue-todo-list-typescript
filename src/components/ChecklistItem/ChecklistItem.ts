// External Dependencies
import { Component, Vue, Prop, Emit } from 'vue-property-decorator';

@Component
export default class Checklist extends Vue {
    // Props
    @Prop() indexKey!: number;
    @Prop() itemTitle!: string;

    // Initial Data
    itemChecked = false;
    editingItem = false;
    itemTitleDone = this.itemTitle;
    itemTitleEdit = this.itemTitle;

    // Methods

    editItem() {
        if (!this.itemChecked) {
            this.editingItem = true;
        }
    }

    cancelEdit() {
        this.editingItem = false;
        this.itemTitleEdit = this.itemTitleDone;
    }

    doneEdit() {
        this.itemTitleDone = this.itemTitleEdit;
        this.editingItem = false;
        if (this.itemTitleDone === '') {
            this.deleteItem();
        }
    }

    @Emit('deleteItem')
    deleteItem() {
        return this.indexKey;
    }
}
