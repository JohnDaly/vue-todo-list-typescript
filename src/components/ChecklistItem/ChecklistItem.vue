<template>
    <b-field>
        <b-checkbox v-model='itemChecked' />
        <span class='item-static' :class="{'item-checked': itemChecked}" v-show='!editingItem' @dblclick='editItem'>
            {{itemTitleDone}}
        </span>
        <b-input v-model='itemTitleEdit' v-show='editingItem' :disabled='itemChecked' @blur="doneEdit"
            @keyup.native.enter="doneEdit" @keyup.native.esc="cancelEdit" expanded />
        <p class='control'>
            <button class='button is-danger' @click='deleteItem'>
                <b-icon icon='close' />
            </button>
        </p>
    </b-field>
</template>

<script lang="ts" src="./ChecklistItem.ts"></script>

<style scoped lang="scss">
.item-static {
  width: 100%;
  width: -moz-available;
  width: -webkit-fill-available;
  width: fill-available;
  display: flex;
  align-items: center;
  border: 1px #dbdbdb solid;
  padding-left: 10px;
}

.item-checked {
  text-decoration: line-through;
}
</style>