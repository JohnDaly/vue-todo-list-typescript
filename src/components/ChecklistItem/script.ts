// External Dependencies
import { Component, Vue, Prop, Emit, Watch } from 'vue-property-decorator';

// Internal Dependencies
import { ChecklistTodo } from '../Checklist/script';

@Component
export default class ChecklistItem extends Vue {
    // Props
    @Prop() todoKey!: string;
    @Prop() title!: string;
    @Prop() completed!: boolean;

    // Initial Data
    itemChecked = false;
    editingItem = false;
    itemTitleDone = this.title;
    itemTitleEdit = this.title;

    // Methods

    /**
     * Sets the value of 'editingItem' to true if the item
     * hasn't already been checked off as completed
     */
    editItem() {
        if (!this.itemChecked) {
            this.editingItem = true;
        }
    }

    /**
     * Reverts the item back to its previous description
     * and hides the editing input field
     */
    cancelEdit() {
        this.editingItem = false;
        this.itemTitleEdit = this.itemTitleDone;
    }

    /**
     * Sets the description of the item to the value of the
     * input field that is used to edit the description. If the user
     * has entered a blank value, the item is deleted
     */
    doneEdit() {
        this.itemTitleDone = this.itemTitleEdit;
        this.editingItem = false;
        if (this.itemTitleDone === '') {
            this.deleteItem();
        }
    }

    /**
     * Emits the 'deleteItem' event, with the value
     * being the indexKey of the this item
     */
    @Emit('deleteItem')
    deleteItem() {
        return this.todoKey;
    }

    /**
     * Emits the 'completedItem' event, with the value
     * being the todoKey of the item
     */
    @Watch('itemChecked')
    @Watch('itemTitleDone')
    @Emit('onChange')
    updateItem(): ChecklistTodo {
        return {
            key: this.todoKey,
            title: this.itemTitleDone,
            completed: this.itemChecked,
        };
    }
}
