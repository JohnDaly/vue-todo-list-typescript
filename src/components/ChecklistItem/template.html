<b-field>
    <!-- A checkbox to toggle whether or not the todo is complete -->
    <b-checkbox v-model='itemChecked' />

    <!-- Shows the text for a todo that is not being edited -->
    <span
        class='item-static'
        :class="{'item-checked': itemChecked}"
        v-show='!editingItem'
        @dblclick='editItem'
    >
        {{itemTitleDone}}
    </span>
    
    <!-- The input field for when a todo is being edited -->
    <b-input
        v-model='itemTitleEdit'
        v-show='editingItem'
        :disabled='itemChecked'
        @blur="doneEdit"
        @keyup.native.enter="doneEdit"
        @keyup.native.esc="cancelEdit"
        expanded
    />
    
    <!-- The button to remove todos from the list -->
    <p class='control'>
        <button class='button is-danger' @click='deleteItem'>
            <b-icon icon='close' />
        </button>
    </p>
</b-field>