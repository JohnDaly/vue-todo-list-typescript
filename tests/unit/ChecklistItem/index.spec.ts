// External Dependencies
import { shallowMount } from '@vue/test-utils';

// Internal Dependencies
import ChecklistItem from '@/components/ChecklistItem/index.vue';

describe('ChecklistItem.vue', () => {
    it('should render properly', () => {
        const wrapper = shallowMount(ChecklistItem, {});
        expect(wrapper).toMatchSnapshot();
    });
});
