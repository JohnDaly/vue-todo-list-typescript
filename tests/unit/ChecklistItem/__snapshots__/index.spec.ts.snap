// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChecklistItem.vue should render properly 1`] = `
<b-field-stub addons="true">
  <b-checkbox-stub truevalue="true"></b-checkbox-stub> <span class="item-static">

    </span>
  <b-input-stub expanded="true" usehtml5validation="true" type="text" hascounter="true" customclass="" style="display: none;"></b-input-stub>
  <p class="control"><button class="button is-danger">
      <b-icon-stub icon="close"></b-icon-stub>
    </button></p>
</b-field-stub>
`;
